nncf/experimental/common/pruning/nodes_grouping.py

```diff
@@ -0,0 +1,466 @@
+    IdentityMaskForwardOps unifies operations that propagate pruning masks as is (relu, swish etc.), whereas
+    Convolution unifies different convolution operations (conv1d, conv2d, conv3d) which accepts some input masks and
+    provide some output masks.
+    :param prune_operations_types: types of operations with prunable parameters.
+    :param dump_dir: path to the directory for dumping debug files.
+    :return: list of groups with parameters of pruning.
+    """
+    # 1. Initialize masks for producing nodes
+    all_nodes_to_prune = graph.get_nodes_by_types(prune_operations_types)  # type: List[NNCFNode]
+    roots = {}
+    for node in all_nodes_to_prune:
+        assert isinstance(node.layer_attributes, (LinearLayerAttributes, ConvolutionLayerAttributes))
+        pruning_dim = node.layer_attributes.get_target_dim_for_compression()
+        output_tensors_shapes = [x.tensor_shape for x in graph.get_output_edges(node)]
+        assert len(output_tensors_shapes) == 1 or len( set(output_tensors_shapes)) <= 1, node.node_name
+        output_tensors_shape = output_tensors_shapes[0]
+        target_output_dim_for_compression = len(output_tensors_shape) - 1
+        root_group = PropagationGroup(
+            blocks=[
+                PropagationBlock(
+                    producer=MaskProducer(node.node_id),
+                    pruning_dimension=pruning_dim
+                )
+            ]
+        )
+        mask = PropagationMask(
+            dim_groups_map={
+                target_output_dim_for_compression: [root_group]
+            }
+        )
+        roots[node.node_id] = root_group
+        node.data['output_mask'] = mask
+
+    def get_attributes_fn(node: NNCFNode) -> Dict[str, Any]:
+        result = {'metatype': str(node.metatype.name), 'node_id': str(node.node_id)}
+        if node.layer_attributes:
+            result.update(map(lambda pair: (pair[0], str(pair[1])), node.layer_attributes.__dict__.items()))
+        if 'output_mask' in node.data:
+            output_mask = node.data['output_mask']
+            if output_mask:
+                result['output_mask'] = str(output_mask)
+        return result
+
+    try:
+        # 2. Propagate masks
+        MaskPropagationAlgorithm(graph, pruning_operations_metatypes).mask_propagation()
+    finally:
+        if dump_dir is not None:
+            block_graph = build_block_hierarchy(list(roots.values()))
+            write_dot_graph(block_graph, dump_dir / 'latest_block_group_hierarchy.dot')
+            save_for_netron(graph, str(dump_dir / 'latest_propagated_graph.xml'), get_attributes_fn=get_attributes_fn)
+
+    # 3. Collect groups from producers
+    blocks_map: Dict[int, List[List[PropagationBlock]]] = {}
+    for id, group in roots.items():
+        blocks_map[id] = group.get_blocks_on_leaves()
+
+    # Filter non closing and duplicated groups
+    pruning_groups = []  # type: List[PruningGroup]
+    finished_producers = []
```

Comment:
`finished_producers` mean seen producers?

Reply:
yes, `selected` might be more clear name